### PR TITLE
Create a Readme for  DP-1.16: Ingress traffic classification and rewrite

### PR DESCRIPTION
**Pull request by ihebboubaker**
This test aims to validate the functionality of ingress traffic classification and subsequent packet remarking (rewrite) on a Device Under Test (DUT). The DUT's configuration will be evaluated against the OpenConfig QOS model, and traffic flows will be analyzed to ensure proper classification, marking, and forwarding.

### CODE REVIEW

**Review by dplore — COMMENTED** [review]
These look very similar to the DP-* tests (DP = dataplane, maybe should've been named QOS for easier searching).  The main difference I see is the use of MPLS EXP.

I think it would make more sense to add a subtest adding the MPLS EXP use case to the appropriate DP-* tests.

**Comment by ihebboubaker** [comment]
Absolutely, but the existing DP tests utilize Weighted Round Robin (WRR) queues, while our tests exclusively employ strict priority queues. This difference in queueing mechanisms necessitates the creation of these new tests.

**Comment by dplore** [comment]
> Absolutely, but the existing DP tests utilize Weighted Round Robin (WRR) queues, while our tests exclusively employ strict priority queues. This difference in queueing mechanisms necessitates the creation of these new tests.

Ah, I overlooked that.  So the DP tests have a mix, where one class is a priority queue and the others are indeed weighted round robin.   

For consistency with the test id's, can you renumber these tests to be DP-* numbers?  Refer to and update the [testregistry.textproto](https://github.com/openconfig/featureprofiles/blob/main/testregistry.textproto) to assign the specific numbers so there is not a conflict.

File: feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+* DUT Configuration:
+    * Configure the DUT's ingress and egress interfaces.
+    * Apply QoS classifiers using the OpenConfig QOS model, matching packets based on DSCP/TC/EXP values as per the classification table.
+    * Configure packet remarking rules based on the marking table.
+
+* Traffic Generation:
+    * Use the traffic generator to send:
+        * IPv4 packets with various DSCP values
+        * IPv6 packets with various TC values
+        * MPLS packets with different EXP values, ensuring scenarios with and without label popping.
```
Comment (sugrimov) [inline feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md:18]: * MPLS traffic with swap action.
* IPv4-over-MPLS traffic with pop action
* IPv6-over-MPLS traffic with pop action
* IPv4-over-MPLS-over-IPv4GRE with GRE decapsulate action and MPLS swap/forward
* IPv6-over-MPLS-over-IPv4GRE with GRE decapsulate action and MPLS swap/forward
* IPv4-over-MPLS-over-IPv4GRE with GRE decapsulate action and MPLS pop and IPv4 forward
* IPv6-over-MPLS-over-IPv4GRE with GRE decapsulate action and MPLS pop and IPv6 forward

File: feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    IPv4 TOS      |       IPv6 TC           |         MPLS EXP        |    Forwarding class
+    ------------- | ----------------------- | ----------------------- | ---------------------
+    0             |      0-7                |          0              |         be1
+    1             |      8-15               |          1              |         af1
+    2             |      16-23              |          2              |         af2
+    3             |      24-31              |          3              |         af3
+    4,5           |      32-47              |          4,5            |         af4
+    6,7           |      48-63              |          6,7            |         nc1
+
+    *   Marking table
```
Comment (sugrimov) [inline feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md:38]: Move "forwarding group" to the beginning of the table.

File: feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+## Summary
+
+This test aims to validate the functionality of ingress traffic classification and subsequent packet remarking (rewrite) on a Device Under Test (DUT). The DUT's configuration will be evaluated against the OpenConfig QOS model, and traffic flows will be analyzed to ensure proper classification, marking, and forwarding.
+
+## Procedure
+
+* DUT Configuration:
+    * Configure the DUT's ingress and egress interfaces.
+    * Apply QoS classifiers using the OpenConfig QOS model, matching packets based on DSCP/TC/EXP values as per the classification table.
+    * Configure packet remarking rules based on the marking table.
```
Comment (sugrimov) [inline feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md:12]: * Set up required static MPLS LSP for testing MPLS swap and pop actions.
* Set up required GRE decapsulation policy-forwarding rules required for GRE decap classification testing.

File: feature/experimental/qos/egress_strict_priority_scheduler_with_bursty_traffic/README.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    IPv4 TOS      |       IPv6 TC           |         MPLS EXP        |    Forwarding class
+    ------------- | ----------------------- | ----------------------- | ---------------------
+    0             |      0-7                |          0              |         be1
+    1             |      8-15               |          1              |         af1
+    2             |      16-23              |          2              |         af2
+    3             |      24-31              |          3              |         af3
+    4,5           |      32-47              |          4,5            |         af4
+    6,7           |      48-63              |          6,7            |         nc1
+
+* Traffic Generation:
```
Comment (sugrimov) [inline feature/experimental/qos/egress_strict_priority_scheduler_with_bursty_traffic/README.md:79]: Generate steady state traffic from ATE Port 1. Take the traffic profile from QOS-1.3

Generate bursty traffic from ATE Port 2 for all traffic classes simultaneously with the same profile for each class:
Frame size: 256
Inter-packet gap: 12
Burst size: 50000
Inter-burst gap: 100ms

Verify that no high priority traffic is dropped (AF3/AF4).

File: feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md (hunk 1 above)
Comment (ihebboubaker) [reply feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md:18]: We are currently focused on testing traffic classification and rewriting. Introducing GRE/MPLS actions might overcomplicate this specific test, as we have separate tests dedicated to evaluating MPLS swap/pop/forward actions and GRE decapsulation.

Comment (dplore) [reply feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md:18]: Seems like the swap and pop should be covered here.  They can be subtests with specific id's specified to help tie together the readme and code.   (separate t.Run's each with a QOS-1.1.x identifier)

It's not a hard rule, but it seems good to add the GRE decap actions to a separate test focused on GRE decap. If anything to separate pass/fail results for this functionality that builds on top of the base MPLS qos features.

Comment (ihebboubaker) [reply feature/experimental/qos/ingress_traffic_classification_and_rewrite/README.md:18]: yes the MPLS actions are covered here by Generating "MPLS packets with different EXP values, ensuring scenarios with and without label popping."
I will change it as sugrimov suggested with : 
* MPLS traffic with swap action.
* IPv4-over-MPLS traffic with pop action.
* IPv6-over-MPLS traffic with pop action.

**Comment by ihebboubaker** [comment]
I am deleting the QOS-1.3 readme file as there is a separate PR https://github.com/openconfig/featureprofiles/pull/3043

File: feature/qos/otg_tests/ingress_traffic_classification_and_rewrite_test/README.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
+# DP-1.16: Ingress traffic classification and rewrite
+
+## Summary
+
+This test aims to validate the functionality of ingress traffic classification and subsequent packet remarking (rewrite) on a Device Under Test (DUT). The DUT's configuration will be evaluated against the OpenConfig QOS model, and traffic flows will be analyzed to ensure proper classification, marking, and forwarding.
+
+## Topology
```
Comment (sugrimov) [inline feature/qos/otg_tests/ingress_traffic_classification_and_rewrite_test/README.md:7]: Please add a link to topology as per the template. 

```
## Testbed type

*  [`featureprofiles/topologies/atedut_2.testbed`](https://github.com/openconfig/featureprofiles/blob/main/topologies/atedut_2.testbed)
```

File: feature/qos/otg_tests/ingress_traffic_classification_and_rewrite_test/README.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    * Configure packet remarking rules based on the marking table.
+    * Configure Static MPLS LSP with MPLS pop and IPv4/IPv6 forward actions for a specific label (ie. 20-100).
+    * Configure Static MPLS LSP MPLS swap/forward actions for a specific label (ie. 101-200).
+
+* Traffic Generation:
+    * Use the traffic generator to send:
+        * IPv4 packets with various DSCP values
+        * IPv6 packets with various TC values
+        * MPLS traffic with swap action.
+        * IPv4-over-MPLS traffic with pop action.
```
Comment (sugrimov) [inline feature/qos/otg_tests/ingress_traffic_classification_and_rewrite_test/README.md:25]: Let's add "MPLS push" action as well for IPv4 and IPv6 traffic.

**Review by sugrimov — APPROVED** [review]

File: feature/qos/otg_tests/ingress_traffic_classification_and_rewrite_test/README.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+This test aims to validate the functionality of ingress traffic classification and subsequent packet remarking (rewrite) on a Device Under Test (DUT). The DUT's configuration will be evaluated against the OpenConfig QOS model, and traffic flows will be analyzed to ensure proper classification, marking, and forwarding.
+
+## Topology
+
+* ATE:port1 <-> port1:DUT:port2 <-> ATE:port2
+
+## Procedure
+
+* DUT Configuration:
```
Comment (dplore) [inline feature/qos/otg_tests/ingress_traffic_classification_and_rewrite_test/README.md:13]: Please add headings in the style of DP-1.16.x for each section as per the [template](https://github.com/openconfig/featureprofiles/blob/main/doc/test-requirements-template.md#procedure).

Each section should read like a 'subtest' (mapping to function in the code with a pass/fail result).  We are only testing the features of interest on the DUT.  The pre-requisite configuration of the test environment (ATE and DUT) should be in the `Test Environment Setup` section.   So you should split out the things like config of interfaces and routing protocols into the `Test Environment Setup` and keep separate from the gnmi.Set for QoS config push and the traffic verifications which can be organized as you have below into DP-1.16.x subtests.

File: feature/qos/otg_tests/ingress_traffic_classification_and_rewrite_test/README.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+rpcs:
+  gnmi:
+    gNMI.Set:
+    gNMI.Subscribe:
+```
+
+## Minimum DUT platform requirement
+
+* MFF - A modular form factor device containing LINECARDs, FABRIC and redundant CONTROLLER_CARD components
```
Comment (sugrimov) [inline feature/qos/otg_tests/ingress_traffic_classification_and_rewrite_test/README.md:153]: As per guidelines we don't need modular chassis for this test.

**Review by sugrimov — APPROVED** [review]

**Review by dplore — APPROVED** [review]
LGTM.  

Agreed with @sugrimov that this does not require MFF device.  I'll merge and then we can update as I am not sure @ihebboubaker is available to make the change.

**Pull request merged**